src/features/fix-all.ts

```diff
@@ -0,0 +1,206 @@
+    readonly fixName: string;
+}
+
+async function buildIndividualFixes(
+    fixes: readonly AutoFix[],
+    client: TspClient,
+    file: string,
+    documents: LspDocuments,
+    diagnostics: readonly lsp.Diagnostic[]
+): Promise<lsp.TextDocumentEdit[]> {
+    const edits: lsp.TextDocumentEdit[] = [];
+    for (const diagnostic of diagnostics) {
+        for (const { codes, fixName } of fixes) {
+            if (!codes.has(diagnostic.code as number)) {
+                continue;
+            }
+
+            const args: tsp.CodeFixRequestArgs = {
+                ...toFileRangeRequestArgs(file, diagnostic.range),
+                errorCodes: [+diagnostic.code!]
+            };
+
+            const response = await client.request(CommandTypes.GetCodeFixes, args);
+            if (response.type !== 'response') {
+                continue;
+            }
+
+            const fix = response.body?.find(fix => fix.fixName === fixName);
+            if (fix) {
+                edits.push(...fix.changes.map(change => toTextDocumentEdit(change, documents)));
+                break;
+            }
+        }
+    }
+    return edits;
+}
+
+async function buildCombinedFix(
+    fixes: readonly AutoFix[],
+    client: TspClient,
+    file: string,
+    documents: LspDocuments,
+    diagnostics: readonly lsp.Diagnostic[]
+): Promise<lsp.TextDocumentEdit[]> {
+    const edits: lsp.TextDocumentEdit[] = [];
+    for (const diagnostic of diagnostics) {
+        for (const { codes, fixName } of fixes) {
+            if (!codes.has(diagnostic.code as number)) {
+                continue;
+            }
+
+            const args: tsp.CodeFixRequestArgs = {
+                ...toFileRangeRequestArgs(file, diagnostic.range),
+                errorCodes: [+diagnostic.code!]
+            };
+
+            const response = await client.request(CommandTypes.GetCodeFixes, args);
+            if (response.type !== 'response' || !response.body?.length) {
+                continue;
+            }
+
+            const fix = response.body?.find(fix => fix.fixName === fixName);
+            if (!fix) {
+                continue;
+            }
+
+            if (!fix.fixId) {
+                edits.push(...fix.changes.map(change => toTextDocumentEdit(change, documents)));
+                return edits;
+            }
+
+            const combinedArgs: tsp.GetCombinedCodeFixRequestArgs = {
+                scope: {
+                    type: 'file',
+                    args: { file }
+                },
+                fixId: fix.fixId
+            };
+
+            const combinedResponse = await client.request(CommandTypes.GetCombinedCodeFix, combinedArgs);
+            if (combinedResponse.type !== 'response' || !combinedResponse.body) {
+                return edits;
+            }
+
+            edits.push(...combinedResponse.body.changes.map(change => toTextDocumentEdit(change, documents)));
+            return edits;
+        }
+    }
+    return edits;
```

Comment:
Just an observation.

The `buildIndividualFixes` and `buildCombinedFix` do a lot of same logic,
wiith the exception that the `buildCombinedFix` has some additional code that fires the `CommandTypes.GetCombinedCodeFix` request, and adds additional text edits... 

The only confusing thing is when is preferable to call `buildIndividualFixes` and when it is preferable to call `buildCombinedFix`?

Reply:
I don't know. This is mostly just code copied from VSCode's typescript language features. They know best what to do since they work on both the language features and typescript.